src/djinterop/enginelibrary/performance_data_format.hpp

```diff
@@ -33,74 +33,118 @@ namespace enginelibrary
 {
 struct beat_data
 {
-    beat_data() noexcept = default;
-
     boost::optional<sampling_info> sampling;
     std::vector<beatgrid_marker> default_beatgrid;
     std::vector<beatgrid_marker> adjusted_beatgrid;
 
+    beat_data() noexcept = default;
+
+    friend bool operator==(
+        const beat_data& first, const beat_data& second) noexcept
+    {
+        return first.sampling == second.sampling &&
+               first.default_beatgrid == second.default_beatgrid &&
+               first.adjusted_beatgrid == second.adjusted_beatgrid;
+    }
+
     std::vector<char> encode() const;
     static beat_data decode(const std::vector<char>& compressed_data);
 };
 
 struct high_res_waveform_data
 {
-    high_res_waveform_data() noexcept = default;
-
     double samples_per_entry;
     std::vector<waveform_entry> waveform;
 
+    high_res_waveform_data() noexcept = default;
+
+    friend bool operator==(
+        const high_res_waveform_data& first,
+        const high_res_waveform_data& second) noexcept
```

Comment:
I tend to use non-friend non-member functions, even when the members of the struct are all public.  Any reason to choose a friend 'member' function here?

Reply:
It has nothing to do with whether or not the members are public. Using a friend function is just the prefered way here because its symmetric.

Reply:
That's not what I mean.  I was wondering why you chose a friend member function definition, as opposed to a non-friend non-member:

```cpp
struct high_res_waveform_data
{
    ...
};

inline bool operator==(
    const high_res_waveform_data& first,
    const high_res_waveform_data& second) noexcept
{
    return first.samples_per_entry == second.samples_per_entry &&
           first.waveform == second.waveform;
}
```

I won't block the PR on this, but the non-friend non-member form has been used elsewhere already in the codebase (see `pad_color.hpp` and `semantic_version.hpp`).

Reply:
Ah sorry. My reason for using fried member functions is because then I don't have to think about whether the members are public or not. It also needs less refactoring when a non-public member is added later. Of course for our simple structs this will not happen, so you can change it if you want to.